modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/model.scala

```diff
@@ -222,7 +257,12 @@ object Queue {
 /**
   * A list of Steps grouped by target and instrument.
   */
-case class Sequence[A](id: String, steps: NonEmptyList[Step[A]])
+case class Sequence[A](
+  id: String,
+  steps: NonEmptyList[Step[A]],
+  // Unrooted executions
```

Comment:
The `Zipper` structure only exists at the very top level, at the `Execution Queue` level. We could make extra zippers at each level but I don't think it'd be worth it since, AFAIK, when executing an *Execution Queue* (which looks like a tree), we need to precisely locate where is the actual execution in reference to the top level structure.

While executing we need to know in which `Sequence` a `Step` is located and in which position in the `Queue` is that `Sequence`. An `Execution` must always belong to a `Step`, a `Step` must always belong to a `Sequence` and so forth. So you can only create the Zipper when all elements in the tree are properly connected.

The `Zipper` in scalaz works for plain `List`s, but it's the same idea for our custom data type.

Actually, [`TreeLoc`](https://oss.sonatype.org/service/local/repositories/releases/archive/org/scalaz/scalaz_2.11/7.2.5/scalaz_2.11-7.2.5-javadoc.jar/!/index.html#scalaz.TreeLoc) which is a `Zipper` for rose trees resembles much closely the zipper for the *Execution Queue*. The crucial difference is that we don't need to move in 4 directions, only forwards and in a limited way backwards when pausing.

Another difference is that our tree is not recursive, we don't need to have an arbitrary number of deepness levels. In our case having a fixed number of levels deep (just 4) enforced by the type system it's beneficial even if a bit cumbersome to encode.

This guarantees that we can't create tree nodes where they don't belong. Having different types at each level guarantees that we can't treat accidentally a `Step` as if it were a `Sequence`, for example.